src/ir_pattern.py

```diff
@@ -0,0 +1,446 @@
+import abc
+from collections import defaultdict
+from dataclasses import dataclass, field
+from itertools import permutations
+from typing import ClassVar, Dict, List, Optional, Type, TypeVar
+
+from .error import static_assert_unreachable
+from .flow_graph import (
+    ArchFlowGraph,
+    BaseNode,
+    FlowGraph,
+    InstrRef,
+    Instruction,
+    LocationRefSetDict,
+    RefSet,
+    Reference,
+    TerminalNode,
+    build_flowgraph,
+)
+from .parse_file import AsmData, Function
+from .parse_instruction import (
+    Argument,
+    AsmAddressMode,
+    AsmGlobalSymbol,
+    AsmInstruction,
+    AsmLiteral,
+    BinOp,
+    Instruction,
+    InstructionMeta,
+    JumpTarget,
+    Location,
+    Macro,
+    RegFormatter,
+    Register,
+    StackLocation,
+    parse_instruction,
+)
+
+
+@dataclass(eq=False, frozen=True)
+class IrPattern(abc.ABC):
+    """
+    Template for defining "IR" patterns that can match against input asm.
+    The matching process uses the FlowGraph and register analysis to compute
+    inter-instruction dependencies, so these patterns can match even when
+    they have been interleaved/reordered by the compiler in the input asm.
+
+    IrPattern subclasses *must* define `parts` and `replacement`, and can
+    optionally implement `check()`.
+
+    For now, the pattern cannot contain any branches, and the replacement
+    must be a single instruction (though, it can be fictive).
+    """
+
+    parts: ClassVar[List[str]]
+    replacement: ClassVar[str]
+
+    flow_graph: FlowGraph
+    replacement_instr: Instruction
+
+    def check(self, m: "IrMatch") -> bool:
+        """Override to perform additional checks/calculations before replacement."""
+        return True
+
+    @classmethod
+    def compile(cls, arch: ArchFlowGraph) -> "IrPattern":
+        missing_meta = InstructionMeta.missing()
+        regf = RegFormatter()
+        replacement_instr = parse_instruction(cls.replacement, missing_meta, arch, regf)
+
+        name = f"__pattern_{cls.__name__}"
+        func = Function(name=name)
+        # Add a fictive nop instruction for each input to the replacement_instr
+        # This acts as a placeholder Reference to represent where the input was set
```

Comment:
Can you assert that these are all unique, here or elsewhere. It's implicitly assumed later on